src/XrdCeph/XrdCephBulkAioRead.cc

```diff
@@ -0,0 +1,157 @@
+#include "XrdCephBulkAioRead.hh"
+
+bulkAioRead::bulkAioRead(librados::IoCtx *ct, logfunc_pointer logwrapper, std::string filename, size_t block_sz) {
+  context = ct;
+  block_size = block_sz;
+  file_name = filename;
+  log_func = logwrapper;
+};
+
+bulkAioRead::~bulkAioRead() {
+  clear();
+};
+
+void bulkAioRead::clear() {
+  for (ReadOpData i: buffers){
+    delete std::get<0>(i);
+    delete std::get<2>(i);
+  }
+  buffers.clear();
+  for (auto const& tup: operations) {
+    librados::ObjectReadOperation *op = std::get<0>(tup.second);
+    librados::AioCompletion *cmpl = std::get<1>(tup.second);
+    cmpl->release();
+    delete op;
+  }
+  operations.clear();
+};
+
+int bulkAioRead::addRequest(std::string objname, char* out_buf, size_t size, off64_t offset) {
+  ceph::bufferlist *bl;
+  int *retval;
+  librados::ObjectReadOperation *rop;
+
+  try {
+    bl = new ceph::bufferlist();
+  } catch (std::bad_alloc&) {
+    log_func((char*)"Can not allocate buffer for read (%lu, %lu)", offset, size);
+    return -ENOMEM;
+  }
+
+  try {
+    retval = new int;
+  } catch (std::bad_alloc&) {
+    log_func((char*)"Can not allocate int for read's retval (%lu, %lu)", offset, size);
+    delete bl;
+    return -ENOMEM;
+  }
+
+  auto tup = operations.find(objname);
+  if (operations.end() == tup) {
+    try {
+      rop = new librados::ObjectReadOperation;
+    } catch (std::bad_alloc&) {
+      log_func((char*)"Can not allocate int for read object op (%lu, %lu)", offset, size);
+      delete bl;
+      delete retval;
+      return -ENOMEM;
+    }
+
+    librados::AioCompletion *cmpl = librados::Rados::aio_create_completion();
+    if (0 == cmpl) {
+      log_func((char*)"Can not create completion for read (%lu, %lu)", offset, size);
+      delete bl;
+      delete retval;
+      delete rop;
+      return -ENOMEM;
+    }
+    operations.insert( {objname, std::make_pair(rop, cmpl)} );
+  } else {
+    rop = std::get<0>(tup->second);
+  }
+  buffers.push_back( std::make_tuple(bl, out_buf, retval) );
+
+  rop->read(offset, size, bl, retval);
+  return 0;
+};
+
+void bulkAioRead::wait_for_complete() {
+  std::string obj_name;
+  librados::AioCompletion* cmpl;
+  librados::ObjectReadOperation* op;
+  for (auto const& tup: operations) {
+    obj_name = tup.first;
+    op = std::get<0>(tup.second);
+    cmpl = std::get<1>(tup.second);
+    context->aio_operate(obj_name, cmpl, op, 0);
+  }
+
+  for (auto const& tup: operations) {
+    cmpl = std::get<1>(tup.second);
+    cmpl->wait_for_complete();
+  }
```

Comment:
I asked whether it was possible to use one completion for several operations, but [the docs indeed say "no"](https://docs.ceph.com/en/latest/rados/api/librados/#asynchronous-io): "For each operation, you must first create a `rados_completion_t`, …"

Reply:
Could you create (and destroy) all the completions here, in this function?

Reply:
Yes, I can. My motivation was the following: once read request is declared (read), user should be able to submit it and get results (i.e. run wait_for_complete). That's why allocation is done in read, to avoid the situation when we executed read but failed to execute wait_for_complete due to the lack of memory. 

I can move completions creation to this method if necessary.